File: elixir/nucleotide-count/nucleotide_count.exs
defmodule NucleotideCount do
  @nucleotides [?A, ?C, ?G, ?T]

  @doc """
  Counts individual nucleotides in a DNA strand.

  ## Examples

  iex> NucleotideCount.count('AATAA', ?A)
  4

  iex> NucleotideCount.count('AATAA', ?T)
  1
  """
  @spec count([char], char) :: non_neg_integer
  def count(strand, nucleotide) do
    Enum.reduce(strand, 0, fn x, a ->
      if x == nucleotide, do: a + 1, else: a
    end)
  end

  @doc """
  Returns a summary of counts by nucleotide.

  ## Examples

  iex> NucleotideCount.histogram('AATAA')
  %{?A => 4, ?T => 1, ?C => 0, ?G => 0}
  """
  @spec histogram([char]) :: map
  def histogram(strand) do
    Enum.reduce(strand, %{?A => 0, ?T => 0, ?C => 0, ?G => 0},
      fn v, a ->
        Map.update!(a, v, fn v ->
          v+1
        end)
      end)
  end
end

File: clojure/nucleotide-count/src/nucleotide_count.clj
(ns nucleotide-count)

(def nucleotide-keys (set "ACTG"))

(defn count-of-nucleotide-in-strand [nucleotide strand]
  {:pre [(nucleotide-keys nucleotide)]}
  (count (filter #(= % nucleotide) strand)))

(defn nucleotide-counts [strand]
  (merge {\A 0 \C 0 \T 0 \G 0} (frequencies strand)))
